blogposts/2020/introducing-codeintel-db.md

```diff
@@ -0,0 +1,65 @@
+
+**TL;DR 1**: If you are a customer currently using precise code intelligence, [skip to the notes below](#notes-for-site-admins) to see how this could affect your instance.
+
+**TL;DR 2**: If you are here for a daily dose of schadenfreude, read about how [we mucked this migration on our production instance](https://eric-fritz.com/articles/migrating-to-postgres/) on Sourcegraph.com.
+
+## The impetus for a second database
+
+A bit over a year ago, the code intelligence backend started as a single node process with access only to its own persistent volume. We chose SQLite as our persistence mechanism for a handful of reasons:
+
+- It was the faster implementation compared to other contending designs we tried.
+- Our MVP did not have access to Postgres at the time, which was owned solely by another team. Only later did we get access to Postgres to store cross-repository metadata and work queue data.
+- SQLite had some nice properties such as easy data warehousing opportunities (on which we have not capitalized) and bulk deletion of expired data was straightforward. All we had to do was delete the SQLite file.
+
+Using SQLite for this data has become an operational and feature burden for the code intelligence team.
+
+1. We need to maintain a periodic process to reconcile the data on disk with the metadata we store in Postgres.
+2. We have good instructions and tips for backup and recovery of our Postgres data store, but have never officially recommended backing up other persistent volumes.
+3. We locked ourselves into a node topology which is not easily scalable. We essentially have a singleton service over a persistent volume. In order to scale this service horizontally (which would become necessary once either the amount of data or the frequency of queries increased), we would need to shard the data across multiple volumes. This throws a lot of wrenches into our plans: we need to be able to automatically move data on increase or decrease of shard counts; we need to be able to route requests to the correct node with that data; we need to ensure that reconciliation between the data on disk and the metadata we store in Postgres is aware that there are multiple nodes where data _could_ reside.
+4. There is no efficient way to implement certain data access patterns we may need in the future. For example, looking for data matching a certain identifier _name_ would require us to open every SQLite database on disk (and, no, _batching it_ is not a solution).
+
+Instead of forging ahead and tackling the complexity we've made for ourselves, we decided to take a step back and re-evaluate whether or not our previous architecture choices were meeting our current needs. And it seems like they did not.
+
+Moving this data into Postgres seems like an obvious choice that will (eventually) be able to reduce or completely remove all of the burdens listed above.
```

Comment:
"Seems like an obvious choice" and "(eventually)" sound wishy-washy and aren't confidence inspiring statements. Are you not confident this was the right decision?

Reply:
Will strengthen!